WordPress/gutenberg
# Nav Block - enable page creation from within Block

<img width="532" alt="Screenshot 2020-01-21 at 13 53 41" src="https://user-images.githubusercontent.com/444434/72810449-7ae5f900-3c55-11ea-9edb-48e3c4a5ae5b.png">


Closes https://github.com/WordPress/gutenberg/issues/18900

This PR aims to allow users to create new Pages (only) on the fly from within the Navigation Block.

### A Note on coupling with WordPress 
Note that the `LinkControl` is within the Block Editor which is supposed to be "unaware" of WordPress. That said the existing implementation in `master` is already coupled to WordPress via `URLInput` which requests data from WP endpoints. Therefore we're following existing precedent here by introducing a prop to pass a handler function create entities on the fly. This is entirely optional and can be configured by the developer so it is less coupled to WP than you might imagine. There is a concept of "creating something" but that's it. We feel this is a good compromise.

## Testing Instructions

* Clear / delete your Posts/Pages.
* Check out PR.
* Create new Page or Post.
* Insert Navigation Block.
* Create a Page
  - start typing anything that could be a valid Page name into the search field
  - see suggestion to "Create Page"
  - click to select and create Page. **Please note:** Link UI will be auto-closed due to [this intentional change](https://github.com/WordPress/gutenberg/commit/06bd2662c8fafacfbd6832aa6017b91915375445).
  - when you reopen the Link UI you should see the Page has been created.
  - verify Page has been created under `Dashboard > Pages`

### Other testing scenarios

* Ensure that "Create" suggestion is not show when entering a URL.
* Try "unusual" values in the input to try and break things (special chars...etc).
* Create a user without permissions to create pages (contributor?). Check that the UI for creating a Page is not exposed to them within the Nav Block.

## Todo

- [x] Consider using `core/data` to create the Pages rather than manual `apiFetch`.
- [x] Fix to ensure Create Page option is displayed if the entry contains spaces (eg: `This is a Test`).
- [x] Implement error handling for when Page cannot be created or times-out.
- [x] ~Implement handling when Page already exists. Perhaps when attempting to create a page which already exists, it seamlessly reuses the existing Page without any need for erroring.~ @marekhrabe and I decided that as per current WP convention a new page should be created with the entered title but with a different slug (eg: `new-page-2`). 

- [x] Handle [a11y concerns regarding the "Create" option not being a valid suggestion which you can move to via the Arrow keys](https://github.com/WordPress/gutenberg/pull/19775#issuecomment-576680517).
- [x] Avoid showing the option to create pages when user doesn't have the capability (`wp.data.select('core').canUser('create', 'pages')`)
- [x] Ensure entity (aka Page creation) is possible using keyboard.
- [x] Implement checks against failure `saveEntityRecord` result types and also `entry` results which are missing appropriate values. This is needed in Nav Link Block `edit` `createEntity()` definition.
- [x] Avoid suggesting URLs for input that isn't a valid URL. This may be better handled in a separate PR as having both options isn't causing any regression. Waiting on https://github.com/WordPress/gutenberg/pull/19871.
- [x] Add "Loading" spinner Icon to `isResolving` state when creating a Page (@jeryj ) - see https://github.com/WordPress/gutenberg/issues/18900#issuecomment-578200190
- [x] Ensure Error notice has correct spacing [as per Design](https://github.com/WordPress/gutenberg/issues/18900#issuecomment-578200190) (@jeryj ).
- [x] Fix failing e2e tests due to use of `uniqueId` invalidating the snapshots. Check with @aduth that using `uniqueId` is a good idea. (@jeryj ).
- [x] Solve [creation of unique IDs for the x2 manually created suggestion types](https://github.com/WordPress/gutenberg/pull/19775#discussion_r374766073).
- [x] Resolve [suggestions label a11y issue](https://github.com/WordPress/gutenberg/pull/19775#discussion_r374805886) (@jeryj).
- [x] Fix additional unwanted spacing below input when no search results. [See here](https://github.com/WordPress/gutenberg/pull/19775#issuecomment-582131560). (@jeryj)
- [x] Write e2e test for Nav Block to cover creating Pages from within Block.
- [x] [Update color references](https://github.com/WordPress/gutenberg/pull/19775#discussion_r375250370) to Sass vars in consultation with Designer (@jeryj ) 
- [x] Fix [component possibly unmounted when state change occurs issue](https://github.com/WordPress/gutenberg/pull/19775#discussion_r374807839).

## How has this been tested?
Unit tests. Manual tests.

## Screenshots 
![Screen Capture on 2020-01-21 at 13-52-09](https://user-images.githubusercontent.com/444434/72810341-4b36f100-3c55-11ea-9b50-b0a66f505447.gif)
## Types of changes
New feature (non-breaking change which adds functionality) 

## Checklist:
